### ISSUE TITLE
Provide KeePassXC as a FlatPak

### ISSUE THREAD

**Issue by iverik-no** [keepassxreboot/keepassxc#1524]
Hello. 

First, thanks for the awesome work you do! 

My question pertains to KeepassXC as Flatpak. I have found KeepassXC on Flathub, but since there's no mention of it on the official KeepassXC page (neither on download, nor blog), I'm hessitant to use it.

https://flathub.org/apps/ (ctrl+F --> "keepassxc")

https://github.com/flathub/org.keepassxc.KeePassXC

I currently use the offical PPA, since the Snaps I've used (KPXC and others) always end up with "Win 95" interface, and a Win 95 window for locating files.

My question is largely this:
* Are these official Flatpaks of KeepassXC?
* ...if no, are there any plans for an official Flatpak?

**Comment by droidmonkey** [comment]
That is NOT official.

**Comment by TheZ3ro** [comment]
This is related to https://github.com/keepassxreboot/keepassxreboot.github.io/issues/28

**Comment by lutracw** [comment]
Slightly confused. droidmonkey says it's not official, but is not the one I found on Flathub the same as the one mentioned on the ticked linked by TheZ3ro, which is there under "offical Flatpak from flathub"? The url for both seems to be the same at least.

**Comment by TheZ3ro** [comment]
@lutracw the first comment is not by a KeePassXC owner.
It's official for flathuib in the sense that flathub directly release it (they maintain it) 
It NOT official by us, in the sense that we don't build it directly ourself and we don't offer support for it

**Comment by lutracw** [comment]
@TheZ3ro I see. Thanks for the clarification. I should think it would be "safe" then, to use it, but I understand that you can't in good conscience vouch for it since you're not the ones to package/provide it. I'll just keep to the PPA for now (already installed, after all), and keeping an eye out for an official Flatpack release by you guys.

**Comment by phoerious** [comment]
It not so much about security, it's more about deployment bugs we cannot fix.

**Comment by rugk** [comment]
Whoever created the flatpak repo, you could just ask them to take it over. I'd also suggest you to do that.

**Comment by TheZ3ro** [comment]
@rugk it's the flathub repository, we can't "take it over". As said in https://github.com/keepassxreboot/keepassxc/issues/1524#issuecomment-368344808 flatpak is an "offical" build since it's compiled locally from github sources, but we do NOT offer support for bug with that distribution channel (unlike snap and appimage)

**Comment by rugk** [comment]
That's stupid. First, most bugs are likely not related to the flatpak and secondly it is kinda unfair to offer it for one technology, but not for the other. To get a widespread software, you should support as much technology stacks as possible.

**Comment by WizardGed** [comment]
They Offer it in 4 Different forms. Source, PPA, Snap, and Appimage.
Demanding a fifth and saying they are "stupid" not to is entitled and
borderline useless when someone else has already packaged and is supporting
it on there own and there is no reason one of the other formats can be used
at essentially no cost.

**Comment by phoerious** [comment]
We simply cannot support all platforms. Bugs that are present in any KeePassXC release will be fixed by us, but any packaging-related bugs have to be fixed by the package provider. We already have enough weird packaging bugs with Snaps and we really don't feel like (and don't have the time for) fixing a whole bunch more for another "bleeding-edge" distribution channel.

**Comment by TheZ3ro** [comment]
Maybe my comment was not clear enough: we do NOT offer support for bugs *in that specific distribution channel*, the guys at flathub will take care of bugs in the KeePassXC flatpak package. 
Bugs in KeePassXC itself are always "supported"*

[*] Note: this project is open-source and maintained on our free-time as an hobby

**Comment by Tids** [comment]
Would be still nice to have you guys taking a look on it too, because at the end of the day the flatpak is [not usable](https://github.com/flathub/org.keepassxc.KeePassXC/issues/4) right now (because of sandboxing).

What I dont understand is that you support PPA and snap, both are Ubuntu packages. 🤔

**Comment by rugk** [comment]
I think https://github.com/flathub/org.keepassxc.KeePassXC/issues/4 is a bug and could be fixed. Maybe the flatpak needs some more permissions or so, or you need to save it somewhere else. So it is not flatpak's problem that this does not work…

**Comment by Tids** [comment]
@rugk could be a Qt/kde runtime thing or a portal bug. I have no idea..

**Comment by droidmonkey** [comment]
Snaps are not just for ubuntu, they can be run on many different distributions. https://snapcraft.io/

In fact based on the icons i see on their website, every single major linux distro...

**Comment by Tids** [comment]
@droidmonkey it's only in the AUR of arch, wich sucks, because the reason for me to use flatpak is to not use the AUR..
it's based on the ubuntu runtime, instead of a nice generic one.
It's not secure, there is no desktop integration yet and there is this annoying visible snap folder in my users home.

So all in all, they can advertise whatever they want, its a Ubuntu thing, while flatpak is the [freedesktop project](https://freedesktop.org/wiki/Software/Flatpak/) project.

**Comment by rugk** [comment]
Also AFAIK it does not have sandboxing on other distros than Ubuntu.

**Comment by RoyiAvital** [comment]
FlatPak is great.
The sand boxing is really a great security feature and matched KeePassXC narrative of securing the user data.

I think the main issue with the FlatPak is the the need to allow access to the `Home` folder.
I wrote about it here - https://github.com/flathub/org.keepassxc.KeePassXC/issues/4#issuecomment-392418043.

**Comment by phoerious** [comment]
AppImage is just a self-contained binary that runs everywhere. Flatpak and Snap both need some other software installed first, so those are not comparable for that matter. The reason we chose Snap was that is was probably the most mature and widely available implementation of confined cross-platform application environments. Generally, we might have chosen Flatpak as well, but it happened to be Snap and we don't want to maintain both.

**Comment by droidmonkey** [comment]
Snaps are 100x easier to define, build, and deploy at this point. Flatpak reminds me of using command line Git... I could not get the Flatpak hosted on Flathub built using the standard commands....

**Comment by droidmonkey** [comment]
There are no performance differences since they run natively in the OS. security is largely dependent on the number of plugs (snaps) or portals (flatpaks) that you connect/open. You can read all about each of them on their respective websites.

**Comment by JaneSmith** [comment]
I'm not involved in this project, so I'll acknowledge that there could be something I'm missing. But as an end user of this software, and as a Linux enthusiast, I find it quite bizarre that a PPA, Snap, and AppImage are provided officially, and yet not Flatpak. Flatpak is becoming the modern standard of application distribution on Linux, whereas Snap is an Ubuntu-specific thing. As someone else mentioned, a PPA is also being official provided, so that's now _two_ things for Ubuntu.

Despite the advertising on the Snap store, it _is_ Ubuntu specific, and if you look at the Linux community (for example, visit Reddit or Hacker News or anywhere where Linux users congregate), you'll see that Flatpak is everyone's go-to rather than Snap. Why? Because it's the standard solution that most everyone is uniting on, a Freedesktop project as somebody mentioned, and is the superior technical solution. Fedora has Flatpak installed by default, and Arch has it in the official repositories. As was mentioned, on Arch, Snap is only available via the Arch User Repository, which is an unsupported way of obtaining unofficial packages, which can be complicated for some people to use, and which can run the risk of malware. Part of the reason of using something like Flatpak or Snap is to get _away_ from the AUR.

There are even at least two distros using Flatpak as the default, primary method of distributing applications: Endless OS and Fedora Silverblue (previously known as Fedora Atomic Workstation). Fedora Silverblue might end up replacing Fedora Workstation at some point in the future. The point is, the world is moving towards Flatpak as the standard distribution model for applications on Linux. By all means, have a PPA, a Snap, or an AppImage if you want. But not providing a Flatpak seems very strange to me.

**Comment by ElCoyote27** [comment]
Would a flatpak help with the issues seen on RHEL7 for v2.5.0 and above? (2.4.3 ran fin but the 2.5.0 AppImage does not even start due to increased requirements in the binary).

**Comment by phoerious** [comment]
The problem is not the packaging format, the problem is the base system used for building that package.

**Comment by 4jNsY6fCVqZv** [comment]
@phoerious What does your build system need to enable reproducible Flatpak builds?

**Comment by 4jNsY6fCVqZv** [comment]
How do you think about that as KeePassXC developers?
https://twitter.com/DanielFore/status/1203071854723731456

**Comment by ovz93br43v7** [comment]
I had installed keepassxc from flathub, do I have to worry or does the current version match the version on Github?

**Comment by droidmonkey** [comment]
The flatpak downloads the sources from GitHub upon building. You always have to worry, but the maintainers of the flatpak have published their build code.

**Comment by ovz93br43v7** [comment]
Danke für deine Antwort. Leider verstehe ich nichts vom programmieren daher kann ich die Modifikationen nicht selber prüfen.  Was meinst du mit "You always have to worry"?

**Comment by wolframroesler** [comment]
I think some official endorsement of https://flathub.org/apps/details/org.keepassxc.KeePassXC by the KeePassXC maintainers and/or a link on https://keepassxc.org/download/#linux (ultimately with some sort of signature/hash to validate authenticity) would make everyone happy, even if it came with a "don't bug us with installation problems" disclaimer. The essential problem at the moment is that people are reluctant to open their password database in a KeePassXC executable not downloaded from an official source because it could do who-knows-what.

**Comment by 4jNsY6fCVqZv** [comment]
I would like to add an ethical quality to the discussion.

Flathub currently does not separate into FLOSS and proprietary apps. If I get even one Free Software app as a flatpak from Flathub, I have the entire remote repository with all proprietary apps in my Software Center.

There will certainly be some users like me who want their KeePassXC Flatpak to come not only from a secure source, as discussed in previous posts, but also from a 100% FLOSS source. 

Therefore I would like to ask you to consider providing an officially maintained Flatpak download from https://keepassxc.org/download/#linux, which is maintained & signed by the KeePassXC team. At least until Flathub offers a libre spin-off for its platform.

**Comment by phoerious** [comment]
Nobody forces you to install any of the proprietary software packages.

**Comment by 4jNsY6fCVqZv** [comment]
Another thing is that my distro wants to go with Flatpak first/only. So I need a secure source for a security sensitive application like your password manager. That's why I care what happens to your Flatpak whenever you want to offer one.

Yes, I do not need to install the proprietary packages, nor do I need to use Flathub. But where should I get it alternatively as Flatpak?
And: Hiding or do not install does not mean that it is no longer there in the remote repo sources.

My concern is not that KeePassXC should not be on Flathub.
Since Flathub does not seem to act actively in the direction to maintain a separate repository for FLOSS packages flathub/flathub#691, I would be glad if you offer an official version as a download e.g. via your website or another safe place as an alternative for people who are looking for 100% Free Software. So my request is about an Flatpak alternative in addition to Flathub. At least for the transition until Flathub also works for users like me.
I know it's extra work and I'm very grateful for what you're doing now!
Nevertheless, my wish remains that you as the developer of KeePassXC not only appreciate my use case, but also actively support my positive FLOSS attitude wherever possible.

What do you think about this?

**Comment by rugk** [comment]
@4jNsY6fCVqZv have not read all of that, but what could solve your use case are [Fedora flatpaks](https://fedoraproject.org/wiki/Workstation/Flatpaks). They are packaged according to the Fedora rules (i.e. FLOSS only e.g.), but can be, of course, installed on any distro.
`registry.fedoraproject.org` is the source for these.

**Comment by 4jNsY6fCVqZv** [comment]
@rugk Hi, thank you for pointing that out! It would certainly be a wonderful alternative if it is FLOSS compatible. At this point it would also be exciting to point out that the developers of the elementary OS platform are already working on such a free repository base for Flatpaks. https://github.com/elementary/flatpak-platform
Nevertheless, the same applies to all alternatives: In order for users to be able to trust them, an official version would be required, which is provided, signed and maintained by KeePassXC.

**Comment by rszibele** [comment]
To put things into perspective, according to https://github.com/paulcarroty/flathub-app-stats/releases/download/20200107122249-08e9356/yearly.json the community run version of KeePassXC on Flathub has been downloaded 85,626 times within the last year.

I also would prefer using the Flatpak, but only if it's an officially supported by the KeePassXC team.

**Comment by phoerious** [comment]
Within a year that's not much. We get three times more direct downloads than that for a single release in two months. I believe the Snap is also way more popular.

**Comment by imchau** [comment]
Flathub is considered trustworthy by the general Linux community. It is easy to confirm at any time that [Flathub is fetching the source tarballs from public repositories and building KeePassXC that way](https://github.com/flathub/org.keepassxc.KeePassXC/blob/master/org.keepassxc.KeePassXC.yml).

Therefore, I feel more confident in the unofficial Flatpak on Flathub to remain safe and unaffected in the case that a malicious party compromises the account of a KeePassXC maintainer.

The daily auto-updating Snap package seems to be entirely in the hands of [@droidmonkey](https://snapcraft.io/keepassxc), while the [PPA is with @phoerious](https://launchpad.net/~phoerious/+archive/ubuntu/keepassxc). When using these ways to acquire a binary of KeePassXC, users place their entire trust in the account safety and goodwill of these individuals.

Additionally, my understanding is that the [prebuilt assets published on Github for every release](https://github.com/keepassxreboot/keepassxc/releases), which are then used for download links on the website and the in-application update checker, are all uploaded by the listed author. Consequently, those binaries could be compromised in a similar manner too. The exception to this are the Source code (zip/tar.gz) links on the bottom such as `https://github.com/keepassxreboot/keepassxc/archive/<tag>.tar.gz`, which, as I understand, are provided directly by Github based on that tagged commit in the repository.

In summary, I believe that the most secure methods to acquire KeePassXC are to personally compile it from source or let a party you already trust by default to do it for you, including Flathub or the repositories of your Linux distribution's official package manger. Am I mistaken to think that the binaries provided by this project's maintainers are more prone to being compromised?

**Comment by phoerious** [comment]
> When using these ways to acquire a binary of KeePassXC, users place their entire trust in the account safety and goodwill of these individuals.

If our accounts or signing keys get compromised, Flathub as a downstream platform will be compromised as well.

**Comment by rugk** [comment]
BTW I personally would prefer flatpak. In contrast to snap, you can also self-host it, so you stay in control, and it is widely supported. (snap is not so nice to setup in many distros and snap's security depends on AppArmor, which is not always available in many distros)
Also – in contrast to snaps – flatpaks do not only claim to be distro-independent, but [actually are](https://kamikazow.wordpress.com/2018/06/08/adoption-of-flatpak-vs-snap-2018-edition/).
Furthermore even [Linux Mint criticizes snap](https://blog.linuxmint.com/?p=3906), because you cannot self-host a snap server as it is proprietary and (thus) also cannot modify the packages that are served centralized by Ubuntu.

**Comment by imchau** [comment]
> > When using these ways to acquire a binary of KeePassXC, users place their entire trust in the account safety and goodwill of these individuals.
> 
> If our accounts or signing keys get compromised, Flathub as a downstream platform will be compromised as well.

Perhaps I should have clarified that, if the public source code itself is maliciously compromised, of course building it downstream would not make a difference. But it would be difficult to hide malicious, compromised source code in a public repository, as it will be easily discovered.

On the other hand, the binaries released by the maintainers' accounts as PPA, Snap, or other binaries on the releases page could contain malicious elements and be built from a different source, while the public code remains clean and unaffected. This is the issue that reproducible builds tackle.

**Comment by phoerious** [comment]
Flatpak downloads the binaries built by us, not the source code.

**Comment by imchau** [comment]
Right now, Flathub is using the manually uploaded source tarball from e.g. `https://github.com/keepassxreboot/keepassxc/releases/download/2.5.4/keepassxc-2.5.4-src.tar.xz` which could potentially contain different code than the repository on Github (which as I noted would be more difficult to covertly compromise). I will open an issue with Flathub to switch the URL to `https://github.com/keepassxreboot/keepassxc/archive/2.5.4.zip` which should is provided by Github and will always contain the contents of the publicly available repository.

**Comment by phoerious** [comment]
It does contain different files than the GitHub-provided one. If you build from the tar.gz, you will need to configure your build differently, otherwise your binary will be marked as a snapshot or pre-release build.

**Comment by imchau** [comment]
> It does contain different files than the GitHub-provided one. If you build from the tar.gz, you will need to configure your build differently, otherwise your binary will be marked as a snapshot or pre-release build.

The maintainer-uploaded release tarballs only include two additional one-liner files over the Github-provided .tar.gz of the same tag.
Only in [.tar.xz]: .gitrev
Only in [.tar.xz]: .version

The rest is identical. If these two files are necessary for a release build, it would be trivial to include them in the Flatpak-specific patchset.

**Comment by kd2flz** [comment]
I would also be _very_ eager to see KeepassXC as an officially supported flatpak. I have used the flatpak, as that is my preferred way to install software, but there are some issues with the user interface, and it didn't receive the update to 2.6.0. I understand this might involve some developer work, but if it were possible that would be AMAZING!

**Comment by darkdragon-001** [comment]
I am currently moving as many GUI applications as possible from PPAs to Flatpak. KeePassXC would be awesome to be part of this, too.

The process of overtaking the flathub repository by the official maintainers is described in the [Flaghub App Submission guide](https://github.com/flathub/flathub/wiki/App-Submission#someone-else-has-put-my-app-on-flathubwhat-do-i-do).

**Comment by rugk** [comment]
@darkdragon-001 KeePassXC is available as an "inofficial" flatpak on Flathub. See https://flathub.org/apps/details/org.keepassxc.KeePassXC. @kd2flz update to v2.6.0 seems to be blocked currently. For more information see/follow https://github.com/flathub/org.keepassxc.KeePassXC/issues/60.

**Comment by arigit** [comment]
FWIW, the version now in flathub is 2.6.0, I have been waiting for 2.6.1 with no luck. It would be nice to have the latest and greatest version in flathub.

Before migrating to flatpak, I tried the keepassxc snap version extensively and it's not there yet, flatpak is more reliable; specially when restricting some capabilities; in my case I wanted the container to not have access to internet. With flatpak this is super simple and works great; with snaps, I couldn't get it to work; the snap version wouldn't honour my dark theme setting, fonts look weird, etc. Flatpak = seamless. All of this in Ubuntu where snaps are supposed to "just work" :)

**Comment by WalnussPower** [comment]
I personnally see Flatpak as a direct competitor too Snap. Given, that a lot of people dislike Canonical and therefore Snap, it seems sensible to offer them the choice.

**Comment by ssokolow** [comment]
> Out of curiosity, what are the difference between FlatPak and Snap security and performance wise?

I know it's been a while but, for anyone else who's wondering:

* Flatpak has no apparent performance downsides aside from possibly a little "having multiple versions of a dependency to ensure ABI stability will increase the chance that a shared library isn't in your disk cache".
* In my experience, snaps are visibly slow to start and are eager to clutter up the `mount` output with a forest of loopback mounts that make `mount` useless in practical terms unless you pipe its output through `grep` to filter them out.

**Comment by soredake** [comment]
> > Out of curiosity, what are the difference between FlatPak and Snap security and performance wise?
> 
> I know it's been a while but, for anyone else who's wondering:
> 
> * Flatpak has no apparent performance downsides aside from possibly a little "having multiple versions of a dependency to ensure ABI stability will increase the chance that a shared library isn't in your disk cache". (Compared to the distro repo approach of rebuilding whatever is necessary to minimize the number of installed versions of the dependencies.)
> * In my experience, snaps are visibly slow to start and are eager to clutter up the `mount` output with a forest of loopback mounts that make `mount` useless in practical terms unless you pipe its output through `grep` to filter them out.

https://github.com/flatpak/flatpak/issues/4187

**Comment by ssokolow** [comment]
Point. I should have phrased it as "I haven't personally observed any performance downsides to using Flatpak".

...and, to be fair, I generally play indie games with low system requirements and already run them under Firejail (which also uses seccomp filtering) to ensure that they can't put non-hidden files/folders into `$HOME` (I'm looking at your use of `getpwuid`/`getpwnam` to try to force making a mess, The Escapists) and to ensure single-player GOG games are *actually* "Whatever experience I play now, I can archive", rather than having some kind of "It crashes if it can't reach the analytics server" bug.

**Comment by soredake** [comment]
https://github.com/flathub/org.keepassxc.KeePassXC/issues/81

**Comment by puelp** [comment]
It's an old post but may be relevant here: https://happyassassin.net/posts/2016/06/16/on-snappy-and-flatpak-business-as-usual-in-the-canonical-propaganda-department/

Basically in that person's opinion, Canonical conducted a heavy PR campaign early on to give off a sense of community consensus surrounding Snap where there was none, to mislead developers into thinking  "Oh snap is backed by Canonical/Ubuntu? Then it must be the dominant Linux standard. OK, let's distribute our software with snap."

So anway...I thought I would add to the arguments in trying to convince the devs to add Flathub support.

First, there is the historical precedent.

Snap is developed by Canonical, while Flatpak and Flathub is developed and backed by a group of people who are primarily employed by Red Hat, and develop for Fedora/Freedesktop/GNOME/Red Hat/systemd/Wayland.

And the historical precedent that I'm speaking of is these guys' software outlasting Canonical's, of which some examples are:

- Unity: competed with GNOME 3 until it was killed and replaced by GNOME 3.
- Mir: competed with Wayland until it was killed and in the process of being replaced by Wayland
- upstart: competed with systemd until it was killed and replaced by systemd
- Ubuntu Software: competed with GNOME Software until it was killed and replaced by a surface-level rebrand of GNOME Software

Basically you can see the picture that I'm trying to paint here. Canonical has a juggernaut in the Ubuntu Linux distribution itself, but when they try to go beyond that and try to make their projects into the community standard, it doesn't work out so well.

In contrast, I do believe that a genuine community consensus is slowly but surely being built up around Flatpak and Flathub.

elementaryOS, Fedora, KDE, Linux Mint, all huge names in the Linux community just off the top of my head, and they have adopted Flathub as a primary method of distributing their apps, while shunning Snap.

For example, you can see that Kdenlive on the Snap Store is outdated (v20.12.3 on the stable channel, v21.04.1 on the beta channel) while the one Flathub is fresh (v21.04.2):
https://snapcraft.io/kdenlive
https://flathub.org/apps/details/org.kde.kdenlive

I hope that the KeePassXC devs are open to changing their minds. :)

**Comment by ghost** [comment]
I'm -1 on an officially supported Flatpak. See: https://drewdevault.com/2021/09/27/Let-distros-do-their-job.html. Historically, when developers have taken over packages, they usually do a very poor job of it because they don't understand the ecosystem or the packaging guidelines. Furthermore, it undermines the community aspect of package management. As long as your software is easy to package or the developer cooperates with downstream, it shouldn't be an issue.

For the security aspect, Flathub is a curated repository and the Flatpaks are built directly from upstream. If you don't trust Flathub, then why would it be any different if a developer was in control? It would be vulnerable to exactly the same issues. None of the countermeasures I've seen either are 100% foolproof against malware and most of it is security theater. It doesn't matter how sandboxed or locked down the application is, at some point the password has to be in memory for it be usable by literally any other software (your browser, the keychain, unattended backups, the list goes on). 

Snap is disliked by a large part of the community for being basically a Canonical initiative with a proprietary backend. Like most Canonical projects, they have a strong not-invented-here syndrome and it is designed to mainly work with Ubuntu. It also is inferior to Flatpak from a technical standpoint that I won't get into, but I think most people prefer Flatpak anyway.

So I find this whole discussion to be a little silly and I vote that we keep things the way that they are, which is to allow the developers on Flathub to keep maintaining the package. They are doing a fine job. If updates are slow, just open an issue and they will get around to it when they get around to it.

**Comment by droidmonkey** [comment]
Snaps and snap definitions have gotten _way worse_ since we first created our snap. However, your argument doesn't hold water. Developers and Distributors need to _work together_ to make the best experience for the end user. It has been demonstrated time and again that when the two don't work together the user experience suffers dramatically. To this end, we will be taking over the FlatPak definition at some point

**Comment by ssokolow** [comment]
Not to mention that systems which provide a consistent runtime across multiple distros are designed with the intent to make insights derived from packaging and testing done by upstream (who have the best understanding of what's being packaged) as broadly applicable to multiple target distros as possible.

In other words, Flatpak runtimes are, by definition, a mechanism to help upstream QA remain applicable despite distro variations.

Traiditional "upstream develops, downstream packages" is intended to allow the package maintenance to be divided among distro experts to make up for variations the upstream doesn't have the knowledge and manpower to anticipate.

This isn't about "Flathub maintainers may be malicious" but more about preventing the next Debian-OpenSSL fiasco. (When Debian package maintainers accidentally rendered OpenSSL server key generation grossly insecure by commenting out some code GCC was warning about that they didn't understand, which was seeding the random number generator.)

**Comment by Hasshu** [comment]
> I'm -1 on an officially supported Flatpak. See: https://drewdevault.com/2021/09/27/Let-distros-do-their-job.html.

@remyabel DeVault fails to understand one simple thing: the shorter the chain of trust is, the better. Instead, he insists on adding more links to it, and that's somehow supposed to make it stronger. I shouldn't have to trust additional people, especially when it comes to security-related software.

> Historically, when developers have taken over packages, they usually do a very poor job of it because they don't understand the ecosystem or the packaging guidelines.

Except that Flatpak is supposed to be universal — there are no different ecosystems or packaging guidelines anymore.

> Flatpaks are built directly from upstream

[Not necessarily](https://github.com/flathub/flathub/wiki/App-Maintenance#strongly-discouraged). "The application on Flathub should reflect, as closely as possible, the application in its unadulterated form, direct from its authors", but "should" =/= "must".

> If you don't trust Flathub, then why would it be any different if a developer was in control?

I would trust Flathub much more if it hosted software packaged by its actual developer(s) instead of random Internet people.

> None of the countermeasures I've seen either are 100% foolproof against malware and most of it is security theater. It doesn't matter how sandboxed or locked down the application is, at some point the password has to be in memory for it be usable by literally any other software (your browser, the keychain, unattended backups, the list goes on).

[Perfect-solution fallacy](https://en.wikipedia.org/wiki/Nirvana_fallacy#Perfect_solution_fallacy).

> If updates are slow, just open an issue and they will get around to it when they get around to it.

Because I have nothing better to do than reminding lazy third parties that they should be doing their job?

**Comment by ssokolow** [comment]
> Not necessarily. "The application on Flathub should reflect, as closely as possible, the application in its unadulterated form, direct from its authors", but "should" =/= "must".

Yeah. Not long ago, having forgotten about that passage, I took the long route to convincing the maintainer of the MPV Flatpak to remove their custom extensions which were overriding my disabling of the OSC GUI to add features I didn't care about.

**Comment by phoerious** [comment]
I couldn't disagree more with that article.

> There are hundreds of Linux distros and each does things differently — the package maintainers are the experts who save you the burden of learning how all of them work.

That is exactly the point. So how do we fix it? Certainly not by fostering more divergence by encouraging distros even more to do their own thing, right? On the contrary: Developers should absolutely ship their own apps. They are the only ones who know their apps inside out and things like backporting CVE patches are a total farce. Nobody knows if a 4-year old backport Frankenstein app is really any more secure than the original unpatched 4-year old app, because who knows if the patch works at all on that old code base? Worst case, it's even less secure because your untested partial patch created a totally new security vulnerability. Nobody tests those things.
Also the argument that packagers take care of creating the best user experience is total bullshit. Yes, they do a great job ironing out certain minor platform bugs, but in the end it is us developers who have to fix such stuff. The number of reports that we have to deal with saying "KeePassXC does this weird thing X on Tipsydibble Linux with Kittybims-i3 wm" is utterly ridiculous.

So developers should package their apps and there should be a good, well-maintained, easy easy-to-use, and universal platform for it. There should of course also be guidelines for basic quality assurance, but they better be universal as well. It is absolutely absurd that even in 2021 it is still near-impossible to make a proper Linux binary. All other operating systems solved that problem decades ago. 

I do agree that Snaps are quite clearly the worst implementation of this model and totally Canonical's NIH syndrome, but at least they try to break with Linux's technical debt of being virtually unpackageable for in their own weird way. If you ask me, I would prefer if all distros switched to Flatpak tomorrow, but as it stands, Flatpak is just yet another package format for us, unfortunately. We have to maintain at least three different Linux package formats already and they are not going away any time soon.

**Comment by Hasshu** [comment]
> Yeah. Not long ago, having forgotten about that passage, I took the long route to convincing the maintainer of the MPV Flatpak to remove their custom extensions which were overriding my disabling of the OSC GUI to add features I didn't care about.

@ssokolow Ah, I've seen that one... Incidentally, mpv [remains](/mpv-player/mpv/issues/6933) the only thing (apart from Nvidia drivers) that forces me to use third-party RPM repos on Fedora. But I digress.

@phoerious Heck, even Linus Torvalds himself [believes it's a problem](https://www.youtube.com/watch?v=Pzl1B7nB9Kc).

**Comment by JaneSmith** [comment]
I just want to chime in and point out that Valve's upcoming new portable Linux PC — the Steam Deck — will be making use of a similar approach to Fedora Silverblue. That is, an immutable base OS with third-party applications installed via modern Flatpaks. Further evidence that Flatpak is the solution most everyone is uniting on and standardising on. Having a modern Flatpak package for this application will make it much easier for users to install on the Steam Deck.

**Comment by ssokolow** [comment]
> See: https://drewdevault.com/2021/09/27/Let-distros-do-their-job.html

I agree with most of what's presented, but I don't agree with the effects. Namely that an elite collection of users (myself included) who know how to build stuff from source should have "This doesn't interest me, therefore it will never get popular" power of veto over projects that may not be targeting their demographic.

> Ship your software as a simple tarball. Don’t ship pre-built binaries


> One thing you shouldn’t do is go around asking distros to add your program to their repos. Once you ship your tarballs, your job is done. It’s the users who will go to their distro and ask for a new package.

How are users supposed to know they want a package if nobody can try it out first?

That's like saying that any music/audio technology which doesn't interest people sufficiently audiophile to know how to hook up a multi-component hi-fi system should be denied a chance to compete in the market. Thus delegitimizing the opinions of anyone outside our/your demographic.

> P.S. Systems which invert this model, e.g. Flatpak, are completely missing the point.

Flathub. Flatpak is just a technology and Red Hat is providing its own APT/RPM-style repo for Fedora that meets the stated maintainership criteria and is pretty obviously intended to become a successor to RPM at the application level for spins like Fedora Silverblue.

**Comment by phoerious** [comment]
>  That is, an immutable base OS with third-party applications installed via modern Flatpaks.

Sounds like classic Android. Funny that most modern Android phones tend to diverge from that pattern again.

**Comment by rugk** [comment]
> Funny that most modern Android phones tend to diverge from that pattern again.

Huh no? Android (respectively it's apps) is/are and always was/were sandboxed. And yes, I guess they got some inspiration from it.
Generally, sandbox-like the desktop systems are worse than mobile platforms, because they miss a lot of strong/solid sandboxing.
See [xckd 1200](https://xkcd.com/1200/). Even flatpak is far from perfect, but it is a try for a larger user base, at least. So yes, that's a very good development.

Anyway, to get on-topic again, don't even know why this issue here is still open, KeePassXC is available as a flatpak already, so what's up here?

**Comment by ssokolow** [comment]
> Even flatpak is far from perfect, but it is a try for a larger user base

To be fair to Flatpak, they are very much focused on retrofitting an improved security baseline onto existing codebases. The more you need to modify the codebases, the harder that is to scale up. Most of their more impressive successes, like XDG portals, were achieved by retrofitting "the platform" in places like the GTK and Qt common dialogs.

> KeePassXC is available as a flatpak already, so what's up here?

It's maintained by others, which adds another opportunity for a mistake or malicious action to open up a hole without someone familiar with the codebase noticing.

**Comment by phoerious** [comment]
> Huh no? Android (respectively it's apps) is/are and always was/were sandboxed.

The apps are sandboxed, but the OS is usually no longer an immutable ROM that needs to be flashed in full in order to update it.

**Comment by gasinvein** [comment]
The KeePassXC flatpak has [several patches](https://github.com/flathub/org.keepassxc.KeePassXC/tree/master/patch/keepassxc) improving quality of life for flatpak users. Maybe those could be merged into KeePassXC upstream codebase? Was it even proposed/considered?

**Comment by droidmonkey** [comment]
And we have now taken over the FlatPak repository and published 2.7.1 to the main line Flathub distribution.

**Comment by ghost** [comment]
Reading your 2.7.1 release blog post, I wonder if you will be announcing the official transition to Flatpak separately?
> we recommend you switch to one of our other two supported Linux packages or to Flatpak once we announce an official Flathub channel.

The blog post gives me the impression that it will still take a while for this transition to be completed. Is that the case?

**Comment by ghost** [comment]
And I also want to take this opportunity to say thank you for making official KeePassXC Flatpak support possible!

**Comment by droidmonkey** [comment]
It's already done, the existing flathub instance is now controlled by us

**Comment by ghost** [comment]
So the updated version her https://flathub.org/apps/details/org.keepassxc.KeePassXC from April 15th is already the official version of the KeePassXC team and therefore safe to install? 

P.s. Would you like to see issues concerning the Flatpak opened here
https://github.com/flathub/org.keepassxc.KeePassXC/issues
or here https://github.com/keepassxreboot/keepassxc/issues?
At least the link on https://flathub.org/apps/details/org.keepassxc.KeePassXC refers to the KeePassXC repository for the issues.

**Comment by droidmonkey** [comment]
If it is a flatpak packaging or sandbox issue it goes in the flathub repo

**Comment by ghost** [comment]
It is an issue with screenshots not being up to date. I think that's part of the packaging, isn't it?

Can you imagine giving me a more clear statement on my first question?

**Comment by droidmonkey** [comment]
> So the updated version her [flathub.org/apps/details/org.keepassxc.KeePassXC](https://flathub.org/apps/details/org.keepassxc.KeePassXC) from April 15th is already the official version of the KeePassXC team and therefore safe to install?

Yes

I'll see if I can get the screenshots updated.

**Comment by ghost** [comment]
thank you very much! yes, the screenshots do not show a dark mode yet, for example.
maybe they can be matched regularly with the ones on your website when there are visible changes or new features.

**Comment by ghost** [comment]
P.s. I had also noticed that I do not get to KeePassXC in the Flathub search for "keepass". Only when I enter at least "keepassx". Maybe you can make the application even easier to find by adjusting the keywords?

**Comment by droidmonkey** [comment]
That seems like a major problem with the search function on flathub to me.

**Comment by fossrob** [comment]
> That seems like a major problem with the search function on flathub to me.

Looks like it's being worked on: https://github.com/flathub/linux-store-frontend/issues/280

https://flathub.vercel.app/apps/search/keepass
